### ISSUE TITLE
[AXE-COCONUT]: Element must have visible text as part of accessible name

### ISSUE THREAD

**Issue by 1Copenut** [department-of-veterans-affairs/va.gov-team#5473]
## Description

**UPDATE:** Bumping this up to a severity 2 item because these links all say "Download" when screen readers focus on the download links. It's nearly impossible to determine what users will download without being able to see the screen.

The VAMC /about page has three links to download PDFs at the end of the page. These three links have `aria-label` attributes that just read **Download**. Screen reader users are left to figure out "Download what?" because the accessible (aria) label is the only thing read. The text on the screen is suppressed. Screen shot attached below.

## Type of Issue
Content - these links are likely being built out in the CMS
_Per @davidconlon, this may be handled from the front end_

## Point of Contact

**VFS Point of Contact:** _Trevor_

## Acceptance Criteria

- [x] Axe-coconut error does not appear on future scans
- [x] Screen reader announces a complete label like "Download 2018 Annual Report PDF"

## Environment

* MacOS
* Chrome
* [axe-coconut browser plugin](https://www.deque.com/axe/axe-for-web/early-release/)
* Endpoints:
  * https://www.va.gov/pittsburgh-health-care/psychology-internships-and-fellowships/
  * https://www.va.gov/pittsburgh-health-care/about/
  * https://www.va.gov/pittsburgh-health-care/programs/pirate-program-for-intensive-residential-aphasia-treatment-education/
  * https://www.va.gov/pittsburgh-health-care/research/safety-security/
  * https://www.va.gov/pittsburgh-health-care/research/research-professionals/
  * https://www.va.gov/pittsburgh-health-care/university-drive-campus-map/
  * https://www.va.gov/pittsburgh-health-care/internships-fellowships/forms/

## Steps to Recreate

1. Enter one of the endpoints in browser
2. Open Chrome dev tools
3. Navigate to the axe-coconut tab, and press the Analyze button
4. Verify there are 3 violations found

## Possible Fixes (optional)

Adopt a tested and approved code pattern for the download links. Prior art here:
* [Add documentation for Download Link](https://github.com/department-of-veterans-affairs/vets-design-system-documentation/issues/152#issuecomment-553127340)
* [Quick Tip: Using the HTML5 Download Attribute](https://webdesign.tutsplus.com/tutorials/quick-tip-using-the-html5-download-attribute--cms-23880)

```html
<a
  class="file-download-with-icon"
  href="/files/2019-08/2018-annual-report_web_508.pdf"
  download="2018 Annual Report.pdf"
  type="application/pdf"
>
  <i aria-hidden="true" class="fas fa-download vads-u-margin--0p5"></i>
  2018 Annual Report <dfn>(<abbr title="Portable Document Format">PDF</abbr>)</dfn>
</a>
```

## WCAG or Vendor Guidance (optional)

* [Understanding Success Criterion 2.5.3: Label in Name](https://www.w3.org/WAI/WCAG21/Understanding/label-in-name.html)

## Screenshots or Trace Logs

![Screen Shot 2020-02-04 at 2.30.51 PM.png](https://images.zenhubusercontent.com/5ac217b74b5806bc2bcd3fc8/115825e9-9859-4b28-bf17-82a082a0c634)

---
![Screen Shot 2020-02-04 at 3.32.48 PM.png](https://images.zenhubusercontent.com/5ac217b74b5806bc2bcd3fc8/70900d66-1e5b-469b-8123-dff8b3c9a97b)

**Comment by davidconlon** [comment]
@1Copenut - Is this an issue for all of VA.gov?

@kevwalsh I think this is something that we would need the CMS team to change first before we can fix our front-end.

**Comment by 1Copenut** [comment]
@davidconlon This issue is decentralized; it depends on how the download link is structured, and how the `aria-label` is structured. We have best practice guidance here: https://github.com/department-of-veterans-affairs/vets-design-system-documentation/issues/152#issuecomment-553127340

It's in a comment at the moment, we're looking for a permanent home for things like this and patterns for handling charts, data tables, etc. being discovered as part of the VAMC audit.

**Comment by mmiddaugh** [comment]
@davidconlon to determine if this is in blobby content...may be handled by CMS team

**Comment by davidconlon** [comment]
This is a universal issue and not in the Blobby WYSIWYG content but the "Link to file or video".
Need to coordinate work on Link to file or video" with CMS team and then, if necessary, we may need to update the front-end.

@mmiddaugh can you coordinate issue/blockage with CMS team?

**Comment by mmiddaugh** [comment]
@ethanteague regarding  #2 in the comment above

After our discussion during grooming this morning, I learned the following from Stan, relevant to
> Content has provided a descriptive name to each pdf.  "Download of Altoona Campus Map", "Download the Altoona 2019 Financial document here", etc.

**Comment by mmiddaugh** [comment]
Addressing #3 in comment above: https://github.com/department-of-veterans-affairs/va.gov-team/issues/8578

This issue is now to address the existing issue of labels on the downloadable items with an approach informed by Stan's info above. Needs estimate once approach is determined.

**Comment by mpelzsherman** [comment]
Our download links should be handled by a common [template](https://github.com/department-of-veterans-affairs/vets-website/blob/master/src/site/paragraphs/downloadable_file.drupal.liquid). This template looks for a string in the graphQL data, e.g. `fieldMedia.entity.image.alt`. Unfortunately, as far as I can tell, either that data isn't being populated in Drupal, or it's not being pulled out of Drupal correctly. I'm going to need some help from the CMS team to figure it out.

**Comment by mpelzsherman** [comment]
Actually I may have a fix for this. I'll push up a PR shortly!

**Comment by mmiddaugh** [comment]
@jenstrickland can you please validate this fix?

**Comment by jenstrickland** [comment]
- Two of the endpoints returned 404s

- One endpoint has a link to a PDF and doesn't use the download PDF link pattern

- Four endpoints still have the issue shown in axe-coconut

Recommended fix enclosed.


**Endpoints:**

**404** 

- https://www.va.gov/pittsburgh-health-care/psychology-internships-and-fellowships/

- https://www.va.gov/pittsburgh-health-care/programs/pirate-program-for-intensive-residential-aphasia-treatment-education/

**Issues persist**

https://www.va.gov/pittsburgh-health-care/about/

![Screen Shot 2020-09-30 at 6 29 24 PM](https://user-images.githubusercontent.com/57469/94747018-01f28180-034c-11eb-90cc-f16cee33e569.png)

<br/><br/>

https://www.va.gov/pittsburgh-health-care/research/safety-security/

![Screen Shot 2020-09-30 at 6 27 25 PM](https://user-images.githubusercontent.com/57469/94746443-b5f30d00-034a-11eb-8b80-b5c4560cafb6.png)

<br/><br/>

https://www.va.gov/pittsburgh-health-care/research/research-professionals/

![Screen Shot 2020-09-30 at 6 38 20 PM](https://user-images.githubusercontent.com/57469/94747092-251d3100-034c-11eb-8f51-0f492a75d708.png)

<br/><br/>

https://www.va.gov/pittsburgh-health-care/university-drive-campus-map/

![Screen Shot 2020-09-30 at 6 39 07 PM](https://user-images.githubusercontent.com/57469/94747166-3ebe7880-034c-11eb-9a35-0bd07d350c43.png)

<br/><br/>

https://www.va.gov/pittsburgh-health-care/internships-fellowships/forms/

There is a PDF link that should use the Download PDF link pattern.

![Screen Shot 2020-09-30 at 6 40 17 PM](https://user-images.githubusercontent.com/57469/94747271-70374400-034c-11eb-981f-44d82982b445.png)

<br/><br/>


**Recommended fix**

Current code
```diff
! <a aria-label="Download 2019 Annual Report" class="file-download-with-icon" target="_blank" href="/files/2020-03/2019-Annual-Report_WEB_508.pdf" download="/files/2020-03/2019-Annual-Report_WEB_508.pdf">
!  <i class="fas fa-download vads-u-margin--0p5"></i>
!  2019 Annual Report (PDF)
</a>
```

Recommended code

Follow the [download PDF link pattern](https://github.com/department-of-veterans-affairs/vets-design-system-documentation/issues/152#issuecomment-553127340) (Not currently in the VA Design System, but [is on the list to get added](https://dsva.slack.com/archives/C0NGDDXME/p1600871794002700?thread_ts=1600871128.002400&cid=C0NGDDXME))
```diff
+ <a type="application/pdf" class="file-download-with-icon" target="_blank" href="/files/2020-03/2019-Annual-Report_WEB_508.pdf" download="2019-Annual-Report_WEB_508.pdf">
+  <i class="fas fa-download vads-u-margin--0p5" aria-hidden="true" role="img"></i>
+  Download 2019 Annual Report <dfn><abbr title="Portable Document Format">PDF</abbr>(1.2<abbr title="Megabytes">MB</abbr>)</dfn>
</a>
```

**Comment by mmiddaugh** [comment]
<details>
<summary>Download links all give visible indicators, as seen on [Psychology Internships and Fellowships](https://www.va.gov/pittsburgh-health-care/work-with-us/internships-fellowships/psychology-internships-and-fellowships/) </summary>

![image](https://user-images.githubusercontent.com/55411834/95259685-cab92000-07e4-11eb-8153-6148b9148f1d.png)

</details>

@jenstrickland this is ready for your validation when you are able

**Comment by jenstrickland** [comment]
@mmiddaugh This is nearly there. Using [the Slack response](https://dsva.slack.com/archives/C0FQSS30V/p1602018636291500?thread_ts=1602017293.286000&cid=C0FQSS30V) here:

Looks good. Under the hood, it would be helpful if we could get some changes. I think this is in the ticket. Is my memory correct that adding the type will be something that happens later?

**Here's what I see online:**

```
<a class="file-download-with-icon" target="_blank" href="/files/2020-03/2019-Annual-Report_WEB_508.pdf" download="/files/2020-03/2019-Annual-Report_WEB_508.pdf">
  <i class="fas fa-download vads-u-margin--0p5" aria-hidden="true" role="img"></i>
  Download 2019 Annual Report (PDF)
</a>
```

**Preferred (to align with [download link pattern](https://github.com/department-of-veterans-affairs/vets-design-system-documentation/issues/152#issuecomment-553127340)):**

```diff
<a class="file-download-with-icon" 
      target="_blank" 
      href="/files/2020-03/2019-Annual-Report_WEB_508.pdf" 
+     download="2019-Annual-Report_WEB_508.pdf"
+     type="application/pdf">
      <i class="fas fa-download vads-u-margin--0p5" aria-hidden="true" role="img"></i>
      Download 2019 Annual Report (PDF)
 </a>
```

**Comment by mpelzsherman** [comment]
I suggest we close this ticket, since the original AC have been met, and open a new ticket to bring the download links into full compliance with the download link pattern.

- Fix the "download" attribute to contain only the file name
- Add a "type" attribute, derived from the file extension
- Add file size to the filename, if available in Drupal. This may require changing the graphql query, and the data may not always be available.

**Comment by jenstrickland** [comment]
At this time, selecting to download `Download Veterans Connect, Issue 6 (PDF)` results in downloading a file named `_files_2019-11_VeteransConnect_Issue7_508.pdf` because the link is coded as `download="/files/2019-11/VeteransConnect_Issue6_508.pdf"`.

@1Copenut wrote this ticket on February 4, and a lot has changed since then. The Acceptance Criteria could be updated to be:

- [X] Axe-coconut error does not appear on future scans
- [X] Screen reader announces a complete label like "Download 2018 Annual Report PDF"
- [ ] Downloaded file name matches the name provided to the user in the download link text
- [ ] `type` attribute is included on the download link
- [ ] Where possible, include the size of the file to be downloaded

Trevor, should we update the acceptance criteria or write new tickets to ensure the output matches the design system standard (on the list of things they'll be adding)?

**Comment by mpelzsherman** [comment]
@jenstrickland The code to fix the original ticket is already in production, so I think writing a new ticket makes more sense from a tracking standpoint.

**Comment by mmiddaugh** [comment]
I agree with @mpelzsherman about creating a new issue for this. It could be tagged as a bug, I think - given the incorrect link. That status will ensure it gets addressed more quickly. We can add that additional AC as well.

**Comment by jenstrickland** [comment]
The journey continues in https://github.com/department-of-veterans-affairs/va.gov-team/issues/14614
